rust/agents/relayer/src/msg/processor.rs

```diff
@@ -0,0 +1,239 @@
+            .set(self.message_leaf_index as i64);
+
+        // Scan until we find next index without delivery confirmation.
+        if self
+            .db
+            .retrieve_leaf_processing_status(self.message_leaf_index)?
+            .is_some()
+        {
+            info!(
+                inbox_name=?self.inbox_contracts.inbox.chain_name(),
+                local_domain=?self.inbox_contracts.inbox.local_domain(),
+                idx=?self.message_leaf_index,
+                "skipping since message_index already in DB");
+            self.message_leaf_index += 1;
+            return Ok(());
+        }
+        let message = if let Some(msg) = self
+            .db
+            .message_by_leaf_index(self.message_leaf_index)?
+            .map(CommittedMessage::try_from)
+            .transpose()?
+        {
+            info!(msg=?msg, "working on msg");
+            msg
+        } else {
+            warn!(
+                "leaf in db without message idx: {}",
+                self.message_leaf_index
+            );
+            // Not clear what the best thing to do here is, but there is seemingly an existing
+            // race wherein an indexer might non-atomically write leaf info to rocksdb across a
+            // few records, so we might see the leaf status above, but not the message contents
+            // here.  For now, optimistically yield and then re-enter the loop in hopes that
+            // the DB is now coherent.
+            // TODO(webbhorn): Why can't we yield here instead of sleep?
+            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+            return Ok(());
+        };
+
+        // Skip if for different inbox.
+        if message.message.destination != self.inbox_contracts.inbox.local_domain() {
+            info!(
+                inbox_name=?self.inbox_contracts.inbox.chain_name(),
+                local_domain=?self.inbox_contracts.inbox.local_domain(),
+                dst=?message.message.destination,
+                msg=?message,
+                "message not for local domain, skipping idx {}", self.message_leaf_index);
+            self.message_leaf_index += 1;
+            return Ok(());
+        }
+
+        // Skip if not whitelisted.
+        if !self.whitelist.msg_matches(&message.message) {
+            info!(
+                inbox_name=?self.inbox_contracts.inbox.chain_name(),
+                local_domain=?self.inbox_contracts.inbox.local_domain(),
+                dst=?message.message.destination,
+                whitelist=?self.whitelist,
+                msg=?message,
+                "message not whitelisted, skipping idx {}", self.message_leaf_index);
+            self.message_leaf_index += 1;
+            return Ok(());
+        }
+
+        // If validator hasn't published checkpoint covering self.message_leaf_index yet, wait
+        // until it has, before forwarding the message to the submitter channel.
+        let mut ckpt;
```

Comment:
Would it be better to return if it is not there and then allow the outer loop to re-run this?